### ISSUE TITLE
ASP.NET web site (not regular project) fails to find gRPC native binaries

### ISSUE THREAD

**Issue by jskeet** [grpc/grpc#12570]
Platform: Windows  
gRPC: 1.6.0  
.NET: 4.6.1  
Prerequisites: VS2017 with web development installed (not just regular ASP.NET)

Repro:

- Start VS2017
- From the File / New menu, pick Web Site (*not* Project), and target .NET 4.6.1 (although I doubt that the version matters much)
- Use "Manage NuGet packages" to add Grpc.Core
- Add a new web form (default.aspx)
- In the codebehind (default.aspx.cs) in the `Page_Load` method, add a single line:

         var channel = new Channel("localhost:8001", ChannelCredentials.Insecure);

- Run the app

It will fail when it can't load the native library, which won't have been copied to the bin directory.

I know very little about "web site" projects, unfortunately - there's no project file, just a packages.config.

**Comment by jskeet** [comment]
I've just tried this with System.Data.Sqlite and seen the same effect. Their advice is "you need the native library on the path". I don't *think* we probe the path at the moment, which may be one option - another would be to allow an environment variable to specify where to find the DLL. Or just not support this scenario, potentially. (Ultimately, the support for native libraries in NuGet packages leaves something to be desired, but we need to work around that as best we can.)

**Comment by pereiren** [comment]
Hi Jon,

I am facing this issue as well. I am working in a Proof of Concept to communicate code from a Website project to .NET Core 2 microservices (yeah I know, messy way but you know how Product Teams work with R&D teams and this is our only possibility to refactor our old code in microservices)

I tried as well to add directly the dlls in Bin folder and I got this:
![website](https://user-images.githubusercontent.com/11976485/32991212-7b3fd7d4-cd37-11e7-93cb-7554a8d85a1d.JPG)

`   en Grpc.Core.Internal.UnmanagedLibrary.GetNativeMethodDelegate[T](String methodName)
   en Grpc.Core.Internal.NativeMethods.GetMethodDelegate[T](UnmanagedLibrary library)
   en Grpc.Core.Internal.NativeMethods..ctor(UnmanagedLibrary library)
   en Grpc.Core.Internal.NativeExtension..ctor()
   en Grpc.Core.Internal.NativeExtension.Get()
   en Grpc.Core.GrpcEnvironment.GrpcNativeInit()
   en Grpc.Core.GrpcEnvironment..ctor()
   en Grpc.Core.GrpcEnvironment.AddRef()
   en Grpc.Core.Server..ctor(IEnumerable`1 options)
   en Grpc.Core.Server..ctor()
   en WebSiteServerPoC.Startup.Configuration(IAppBuilder app) en C:\Users\Pereira\Documents\Visual Studio 2017\Projects\WebSiteServerPoC\App_Code\Startup.cs:línea 23
`
Hope it can hlep, I will try to find a solution anyway.

**Comment by jskeet** [comment]
The workaround I'm aware of (assuming it's actually the same error - it's not immediately clear that it is) is to write code to copy the plugin into the bin directory before you make any gRPC calls.

But it sounds like there's probably nothing we can do about the nuget installation for the package not doing this itself - I suspect that Website projects just don't perform the right steps here :(

**Comment by pereiren** [comment]
I tried older versions because I saw this issue noted on last version and it fails on the same way. 

Did you test it specifying an environment variable as you sugggested?  If not, I can give it a try during next days.

Thank you Jon, always a pleasure

**Comment by jskeet** [comment]
No, I haven't tried modifying the path.

**Comment by pereiren** [comment]
It fails also trying to load the native DLLs on path, it is the same that trying to copy them to Bin folder.

I will give RabbitMQ a try to communicate our website project with .NET Core 2 microservices. I wanted to use gRPC but it can't be possible with this kind of website projects.

Anyway thank you Jon for the help and for finding out this issue.

**Comment by pereiren** [comment]
Finally I solved this issue. Instead of using the Nuget package, you need to add the DLL as a reference. It works now.

**Comment by sdg002** [comment]
> Finally I solved this issue. Instead of using the Nuget package, you need to add the DLL as a reference. It works now.

Thanks for the tip. I am facing a similar problem. I have a .NET Framework EXE and a .NET Standard class library that does some Google NLP stuff.
What was the location of the DLL from where you added the referece? In my case Nuget packages are all at the %USERPROFILE% folder.

**Comment by jskeet** [comment]
@sdg002: For most applications you shouldn't normally run into this problem (this issue is specific to ASP.NET Web Site projects), but there can be a transitive dependency issue. If you just add a direct reference to `Grpc.Core` in either of your projects, I'd expect that to solve the problem.

If it doesn't, could you create a new issue with a small example so I can reproduce it? Feel free to add that in the [google-cloud-dotnet](https://github.com/googleapis/google-cloud-dotnet) repo or here - but if it's here please tag me (just include @jskeet in the description) so I can see be notified.

**Comment by steveoh** [comment]
@jskeet https://github.com/aspnet/AspNetCore/issues/12101 this seems related? I tried adding `Grpc.Core` as a dependency and it didn't make a difference.

**Comment by jskeet** [comment]
Not sure - but the gRPC team is likely to be better placed to diagnose it than I am.

**Comment by sreejit-a** [comment]
I am facing a similar issue when trying to use gRPC client in an ASP.Net website. The exact error message I receive is 

```
---> Error loading native library. 
   Not found in any of the possible locations: 
   C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\grpc_csharp_ext.x64.dll,
   C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\runtimes/win/native\grpc_csharp_ext.x64.dll,C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\../..\runtimes/win/native\grpc_csharp_ext.x64.dll (See inner exception for details.) (See inner exception for details.) ---> Autofac.Core.DependencyResolutionException: An exception was thrown while invoking the constructor 'Void .ctor(Microsoft.Extensions.Logging.ILoggerFactory, Microsoft.Xrm.RemotePlugin.Client.Config.ISandboxClientSettings)' on type 'SandboxFabricGrpcClient'. ---> Error loading native library. Not found in any of the possible locations: C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\grpc_csharp_ext.x64.dll,C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\runtimes/win/native\grpc_csharp_ext.x64.dll,C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\../..\runtimes/win/native\grpc_csharp_ext.x64.dll (See inner exception for details.) ---> System.IO.FileNotFoundException: Error loading native library. Not found in any of the possible locations: C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\grpc_csharp_ext.x64.dll,C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\runtimes/win/native\grpc_csharp_ext.x64.dll,C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Temporary ASP.NET Files\root\85effa84\2238b786\assembly\dl3\7dce0beb\00f66c13_3e74d501\../..\runtimes/win/native\grpc_csharp_ext.x64.dll
   at Grpc.Core.Internal.UnmanagedLibrary.FirstValidLibraryPath(String[] libraryPathAlternatives)
   at Grpc.Core.Internal.UnmanagedLibrary..ctor(String[] libraryPathAlternatives)
   at Grpc.Core.Internal.NativeExtension.LoadUnmanagedLibrary()
   at Grpc.Core.Internal.NativeExtension.LoadNativeMethods()
   at Grpc.Core.Internal.NativeExtension..ctor()
   at Grpc.Core.Internal.NativeExtension.Get()
   at Grpc.Core.Internal.NativeMethods.Get()
   at Grpc.Core.GrpcEnvironment.GrpcNativeInit()
   at Grpc.Core.GrpcEnvironment..ctor()
   at Grpc.Core.GrpcEnvironment.AddRef()
   at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials, IEnumerable`1 options)
   at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials)
```

The problem seems to be that ASP.Net copies Grpc.Core dll to Temporary ASP.Net file folders (as part of Dynamic Compilation?) and Grpc.Core looks for native libraries within this assembly location - [NativeExtension.cs](https://github.com/grpc/grpc/blob/2780136fcf9bd268aaddea74b115fbc2faa97c61/src/csharp/Grpc.Core/Internal/NativeExtension.cs#L87) which is not present, effectively blocking ASP.Net apps as a possible gRPC client. 

@jskeet, @steveoh - Wanted to check if there's anyway to overcome this issue or was it solved by other means.

**Comment by jskeet** [comment]
I've had it worked with regular ASP.NET in the past - assuming this isn't a "web site" project, I suggest you open a new issue with your *precise* setup so that the gRPC team can reproduce it.

**Comment by sreejit-a** [comment]
We figured out the issue eventually - We were incorrectlly referring netstandard version of the grpc dlls and not .Net Framework ones - so GetAssemblyPath()  in NativeExtension.cs was returning the actual file path rather than the EscapedCodeBase path.

**Comment by steveoh** [comment]
@sreejit-a we as in the google package owner of the grpc? So this will be fixed when you push an update?

**Comment by apolcyn** [comment]
@jtattermusch have you been able to look into this?

**Comment by jtattermusch** [comment]
We've never got to fix this, but given that VS2017 is quite old now and ASP.NET core (where things work fine) is now the standard, I'm going to close this issue as "Obsolete".

**Comment by steveoh** [comment]
I wasn't using 2017 when this issue affected me. I believe it was 2019. I've since removed the grpc dependency to be able to deploy my application which is not ideal. I would love to see a fix or some attempt at one.

**Comment by jtattermusch** [comment]
@steveoh  can you provide an example project that reproduces the problem? (ideally as a git repository).

**Comment by steveoh** [comment]
It was this project for me. https://github.com/agrc/parole-and-probation/blob/master/src/app.csproj

**Comment by jtattermusch** [comment]
@steveoh  the example project you provided doesn't seem to reference grpc nuget at all.

**Comment by steveoh** [comment]
It’s via the serilog google cloud nuget package. It has a reference to the grpc bits.
